File: swift/334_IncreasingTripletSubsequence.swift
class Solution {
    func increasingTriplet(_ nums: [Int]) -> Bool {
        if nums.isEmpty { return false }
        
        var stacks = [[nums[0]]]
        for i in 1..<nums.count {
            var n = nums[i]
            for j in 0..<stacks.count {
                var stackLength = stacks[j].count
                if n < stacks[j].last! && (stacks[j].count == 1 || stacks[j][0] < n) {
                    stacks[j][stackLength - 1] = n
                    break
                } else if n < stacks[j][0] && j == stacks.count - 1 {
                    stacks.append([n])
                    break
                } else if n > stacks[j].last! {
                    stacks[j].append(n)
                    if stacks[j].count == 3 { return true }
                    break
                }
            }
        }
        return false;
    }
}

File: cpp/334_IncreasingTripletSubsequence.cpp
class Solution {
public:
    bool increasingTriplet(vector<int>& nums) {
        vector<vector<int>> stacks = {{nums[0]}};
        for (int i = 1; i < nums.size(); ++i) {
            int n = nums[i];
            for (int j = 0; j < stacks.size(); ++j) {
                vector<int> &stack = stacks[j];
                if (n < stack.back() && (stack.size() == 1 || n > stack[0])) {
                    stack[stack.size() - 1] = n;
                    break;
                } else if (n < stack[0] && j == stacks.size() - 1) {
                    stacks.push_back({n});
                    break;
                } else if (n > stack.back()) {
                    stack.push_back(n);
                    if (stack.size() == 3) return true;
                    break;
                }
            }
        }
        return false;
    }
};
